packages/client/wallets/aa/src/blockchain/wallets/passkey.ts

```diff
@@ -0,0 +1,111 @@
+    return (params.walletConfig.signer as PasskeySigner).type === "PASSKEY";
+}
+
+type PasskeyValidator = KernelValidator<EntryPoint, "WebAuthnValidator"> & {
+    getSerializedData: () => string;
+};
+
+export class PasskeyWalletService {
+    constructor(private readonly crossmintService: CrossmintWalletService) {}
+
+    public async getOrCreate({ userIdentifier, chain, publicClient, walletConfig, entrypoint }: PasskeyWalletParams) {
+        const validator = await this.getOrCreateSigner({
+            userIdentifier,
+            entrypoint,
+            publicClient,
+            signer: walletConfig.signer,
+        });
+        const kernelAccount = await createKernelAccount(publicClient, {
+            plugins: { sudo: validator },
+            entryPoint: entrypoint.address,
+        });
+
+        await this.crossmintService.storeAbstractWallet({
+            userIdentifier,
+            type: ZERO_DEV_TYPE,
+            smartContractWalletAddress: kernelAccount.address,
+            signerData: this.getSignerData(validator, walletConfig.signer.passkeyName),
+            version: CURRENT_VERSION,
+            baseLayer: "evm",
+            chainId: blockchainToChainId(chain),
+            entryPointVersion: entrypoint.version,
+        });
+
+        return new EVMSmartWallet(this.crossmintService, kernelAccount, publicClient, chain);
+    }
+
+    private async getOrCreateSigner({
+        userIdentifier,
+        entrypoint,
+        publicClient,
+        signer,
+    }: {
+        userIdentifier: UserIdentifier;
+        entrypoint: EntryPointDetails;
+        publicClient: PublicClient;
+        signer: PasskeySigner;
+    }): Promise<PasskeyValidator> {
+        const serializedData = await this.fetchSerializedSigner(userIdentifier);
+        if (serializedData != null) {
+            return deserializePasskeyValidator(publicClient, {
+                serializedData,
+                entryPoint: entrypoint.address,
+            });
+        }
+
+        return createPasskeyValidator(publicClient, {
+            passkeyServerUrl: this.crossmintService.getPasskeyServerUrl(userIdentifier),
+            entryPoint: entrypoint.address,
+            passkeyName: signer.passkeyName,
+            credentials: "omit",
+        });
+    }
+
+    private async fetchSerializedSigner(user: UserIdentifier): Promise<string | null> {
+        const signer = await this.crossmintService.getPasskeyValidatorSigner(user);
+        if (signer == null) {
+            return null;
+        }
+
+        if (signer.type !== "passkeys") {
+            throw new Error("Admin Mismatch"); // TODO custom error as defined within SDK spec
+        }
+
+        return serializePasskeyValidatorData(signer);
+    }
+
+    private getSignerData(validator: PasskeyValidator, passkeyName: string): PasskeySignerData {
+        const fields = deserializePasskeyValidatorData(validator.getSerializedData());
```

Comment:
In an SDK update ZD has improved how this works iirc, so we should be able to cleanup. The upgrade will be non-trivial for other reasons, so leaving what feels like excess (de)serialization here for now.